device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/InternalClient.java

```diff
@@ -184,6 +184,33 @@ public void sendEventAsync(Message message, IotHubEventCallback callback, Object
         deviceIO.sendEventAsync(message, callback, callbackContext, this.config.getDeviceId());
     }
 
+    /**
+     * ASynchronously sends a batch of messages to the IoT Hub
+     * HTTPS messages will be sent in a single batch and MQTT and AMQP messages will be sent individually.
+     * Maximum payload size for HTTPS is 255KB
+     *
+     * @param messages the list of message to be sent.
+     * @param callback the callback to be invoked when a response is received.
+     * Can be {@code null}.
+     * @param callbackContext a context to be passed to the callback. Can be
+     * {@code null} if no callback is provided.
+     *
+     * @throws IllegalArgumentException if the message provided is {@code null}.
+     * @throws IllegalStateException if the client has not been opened yet or is
+     * already closed.
+     */
+    public void sendEventBatchAsync(Set<Message> messages, IotHubEventCallback callback, Object callbackContext)
```

Comment:
Does this have to be a set? What about a more generic type like Collection or Iterable?

Reply:
That is an excellent question. I don't think it has to necessarily be a Set, I tried to follow the other APIs we have in the internal client and used Sets. is there a drawback in using Sets here?

Reply:
I made it into a List to maintain the orders in the collection as Abhipsa mentioned.

Reply:
As long as they need to be ordered, then a List should be fine. In general, if the API can take a more generic collection type, then it should. That way users who like sets can pass in a set, and users who like lists can pass in a list. But that's all moot if we have to maintain order here.